util/dns_resolve.cc

```diff
@@ -4,126 +4,88 @@
-
-#ifdef USE_GAI_ASYNC
-  struct gaicb gai_req;
-  gai_req.ar_name = dns;
-  gai_req.ar_request = &hints;
-  gai_req.ar_result = nullptr;
-  gai_req.ar_service = nullptr;
-
-  struct gaicb* gai_list[] = {&gai_req};
-
-  NotifyStruct* ns = new NotifyStruct;
-
-  struct sigevent sigev;
-  sigev.sigev_notify = SIGEV_THREAD;
-  sigev.sigev_notify_function = DnsResolveNotify;
-  sigev.sigev_value.sival_ptr = ns;
-  sigev.sigev_notify_attributes = nullptr;
-
-  CHECK_EQ(0, getaddrinfo_a(GAI_NOWAIT, gai_list, 1, &sigev));
-  ns->evc.await([ns] { return ns->gate.load(memory_order_relaxed); });
-  delete ns;
-
-  int res = gai_error(&gai_req);
-  if (res == EAI_CANCELED) {
-    return make_error_code(errc::operation_canceled);
+void DnsResolveCallback(void* ares_arg, int status, int timeouts, hostent* hostent) {
+  CHECK(ares_arg != nullptr);
+  auto* cb_args = static_cast<DnsResolveCallbackArgs*>(ares_arg);
+  if (status != ARES_SUCCESS || hostent == nullptr) {
+    cb_args->promise.set_value(make_error_code(errc::address_not_available));
+    return;
   }
 
-  if (res == EAI_NONAME) {
-    return make_error_code(errc::bad_address);
+  if (hostent->h_addrtype != AF_INET) {
+    // We currently only support IPv4
+    cb_args->promise.set_value(make_error_code(errc::address_not_available));
   }
 
-  CHECK_EQ(0, res);
-  servinfo = gai_req.ar_result;
-#else
-  int res = getaddrinfo(dns, NULL, &hints, &servinfo);
-  if (res != 0) {
-    LOG(ERROR) << "Error calling getaddrinfo " << gai_strerror(res);
-    return make_error_code(errc::address_not_available);
+  char** addr = hostent->h_addr_list;
+  if (addr == nullptr) {
+    cb_args->promise.set_value(make_error_code(errc::address_not_available));
+    return;
   }
 
-#endif
+  ares_inet_ntop(AF_INET, *addr, cb_args->dest_ip, INET_ADDRSTRLEN);
+  cb_args->promise.set_value(error_code());
+}
+
+void WaitForChannel(ares_channel channel) {
+  while (true) {
+    fd_set readers, writers;
+    FD_ZERO(&readers);
+    FD_ZERO(&writers);
 
-  res = EAI_FAMILY;
-  for (addrinfo* p = servinfo; p != NULL; p = p->ai_next) {
-    if (p->ai_family == AF_INET) {
-      struct sockaddr_in* ipv4 = (struct sockaddr_in*)p->ai_addr;
-      const char* inet_res = inet_ntop(p->ai_family, &ipv4->sin_addr, dest_ip, INET6_ADDRSTRLEN);
-      CHECK_NOTNULL(inet_res);
-      res = 0;
+    int nfds = ares_fds(channel, &readers, &writers);
+    if (nfds == 0) {
       break;
     }
-    LOG(WARNING) << "Only IPv4 is supported";
+
+    if (select(nfds, &readers, &writers, nullptr, nullptr) == 0) {
+      boost::this_fiber::yield();
+      continue;
+    }
+
+    ares_process(channel, &readers, &writers);
   }
+}
+
+}  // namespace
+
+error_code DnsResolve(const char* dns, uint32_t wait_ms, char dest_ip[]) {
+  ares_channel channel;
+  CHECK_EQ(ares_init(&channel), ARES_SUCCESS);
+
+  DnsResolveCallbackArgs cb_args;
+  cb_args.dest_ip = dest_ip;
+  boost::fibers::future<std::error_code> future = cb_args.promise.get_future();
```

Comment:
no need, just return `cb_args.result` at the end. Since you have this barrier `WaitForChannel` you already take care of resolution, and this is why using futures is redundant here